test/PoolManager.t.sol

```diff
@@ -1193,6 +1210,143 @@ contract PoolManagerTest is Test, Deployers, TokenFixture, GasSnapshot {
+            )
+        );
+
+        vm.etch(hookAddr, vm.getDeployedCode("NoOpTestHooks.sol:NoOpTestHooks"));
+
+        PoolKey memory key =
+            PoolKey({currency0: currency0, currency1: currency1, fee: 100, hooks: IHooks(hookAddr), tickSpacing: 10});
+        manager.initialize(key, SQRT_RATIO_1_1, ZERO_BYTES);
+
+        // Test add liquidity
+        snapStart("modify position with noop");
+        BalanceDelta delta =
+            modifyPositionRouter.modifyPosition(key, IPoolManager.ModifyPositionParams(-120, 120, 10 ether), ZERO_BYTES);
+        snapEnd();
+
+        // Swap
+        IPoolManager.SwapParams memory swapParams =
+            IPoolManager.SwapParams({zeroForOne: true, amountSpecified: 100, sqrtPriceLimitX96: SQRT_RATIO_1_2});
+
+        PoolSwapTest.TestSettings memory testSettings =
+            PoolSwapTest.TestSettings({withdrawTokens: true, settleUsingTransfer: true});
+
+        snapStart("swap with noop");
+        delta = swapRouter.swap(key, swapParams, testSettings, ZERO_BYTES);
+        snapEnd();
+
+        // Donate
+        snapStart("donate with noop");
+        delta = donateRouter.donate(key, 0, 0, ZERO_BYTES);
+        snapStart("donate with noop");
+    }
+
+    function test_noop_succeedsOnAllActions(uint160 sqrtPriceX96) public {
+        // Assumptions tested in Pool.t.sol
+        vm.assume(sqrtPriceX96 >= TickMath.MIN_SQRT_RATIO);
+        vm.assume(sqrtPriceX96 < TickMath.MAX_SQRT_RATIO);
+
+        address payable hookAddr = payable(
+            address(
+                uint160(
+                    Hooks.BEFORE_MODIFY_POSITION_FLAG | Hooks.BEFORE_SWAP_FLAG | Hooks.BEFORE_DONATE_FLAG
+                        | Hooks.NO_OP_FLAG
+                )
+            )
+        );
+
+        vm.etch(hookAddr, vm.getDeployedCode("NoOpTestHooks.sol:NoOpTestHooks"));
+
+        PoolKey memory key =
+            PoolKey({currency0: currency0, currency1: currency1, fee: 100, hooks: IHooks(hookAddr), tickSpacing: 10});
+
+        manager.initialize(key, SQRT_RATIO_1_1, ZERO_BYTES);
+
+        // Test add liquidity
+        assertEq(manager.reservesOf(currency0), 0);
+        assertEq(manager.reservesOf(currency1), 0);
+        BalanceDelta delta =
+            modifyPositionRouter.modifyPosition(key, IPoolManager.ModifyPositionParams(-120, 120, 10 ether), ZERO_BYTES);
+
+        assertEq(delta.amount0(), 0);
+        assertEq(delta.amount0(), 0);
+        assertEq(manager.reservesOf(currency0), 0);
+        assertEq(manager.reservesOf(currency1), 0);
+
+        // Swap
+        IPoolManager.SwapParams memory swapParams =
+            IPoolManager.SwapParams({zeroForOne: true, amountSpecified: 100, sqrtPriceLimitX96: SQRT_RATIO_1_2});
+
+        PoolSwapTest.TestSettings memory testSettings =
+            PoolSwapTest.TestSettings({withdrawTokens: true, settleUsingTransfer: true});
+
+        delta = swapRouter.swap(key, swapParams, testSettings, ZERO_BYTES);
+
+        assertEq(delta.amount0(), 0);
```

Comment:
I think delta will be 0 even if NoOp doesn't work here because there is no underlying liquidity in the pool. (It NoOps the add liquidity before). 

So I think we should check that delta is 0 and there are reserves in the pool manager

Reply:
> I think delta will be 0 even if NoOp doesn't work here because there is no underlying liquidity in the pool. (It NoOps the add liquidity before).

Thats true - it would actually revert in the `swapRouter` for that because it checks if noop was allowed when delta was 0. So I'll just remove these checks - its handled in the other contract. (And reserves are checked on the next line)